Tools/cases_generator/optimizer_generator.py

```diff
@@ -111,19 +122,88 @@ def goto_label(self, goto: Token, label: Token, storage: Storage) -> None:
         self.out.emit(goto)
         self.out.emit(label)
 
+
+class OptimizerConstantEmitter(OptimizerEmitter):
+    def __init__(self, out: CWriter, labels: dict[str, Label], uop: Uop):
+        super().__init__(out, labels)
+        # Replace all outputs to point to their stackref versions.
+        overrides = {
+            outp.name: self.emit_stackref_override for outp in uop.stack.outputs
+        }
+        self._replacers = {**self._replacers, **overrides}
+
+    def emit_stackref_override(
+        self,
+        tkn: Token,
+        tkn_iter: TokenIterator,
+        uop: CodeSection,
+        storage: Storage,
+        inst: Instruction | None,
+    ) -> bool:
+        self.out.emit(tkn)
+        self.out.emit("_stackref")
+        return True
+
+def write_uop_pure_evaluation_region_header(
+    uop: Uop,
+    out: CWriter,
+    stack: Stack,
+) -> None:
+    emitter = OptimizerConstantEmitter(out, {}, uop)
+    emitter.emit("if (\n")
+    assert len(uop.stack.inputs) > 0, "Pure operations must have at least 1 input"
+    for inp in uop.stack.inputs:
+        emitter.emit(f"sym_is_const(ctx, {inp.name}) &&\n")
+    emitter.emit("1) {\n")
```

Comment:
Again, not sure how much we care about how the generated code is formatted, but we could format this a bit nicer and get rid of the trailing `1` perhaps?
```suggestion
    for inp in uop.stack.inputs[:-1]:
        emitter.emit(f"sym_is_const(ctx, {inp.name}) &&\n")
    emitter.emit(f"sym_is_const(ctx, {uop.stack.inputs[-1].name})\n")
    emitter.emit(') {\n')
```
Compare
```c
            if (
                sym_is_const(ctx, left) &&
                sym_is_const(ctx, right) &&
            1) {
```
vs:
```c
            if (
                sym_is_const(ctx, left) &&
                sym_is_const(ctx, right)
            ) {
```